core/drivers/ls_dspi.c

```diff
@@ -0,0 +1,450 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2021 NXP
+ *
+ * DSPI Driver for DSPI Controller
+ *
+ */
+
+#include <assert.h>
+#include <drivers/ls_dspi.h>
+#include <io.h>
+#include <kernel/boot.h>
+#include <kernel/delay.h>
+#include <kernel/dt.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <util.h>
+
+#define NXP_DSPI_COMPATIBLE "fsl,lx2160a-dspi"
+
+/*
+ * Calculate the divide scaler value between expected SCK frequency
+ * and input clk frequency
+ */
+static TEE_Result dspi_convert_hz_to_baud(unsigned int *req_pbr,
+					  unsigned int *req_br,
+					  unsigned int speed_hz,
+					  unsigned int clkrate)
+{
+	/* Valid pre-scaler values for baud rate*/
+	static const unsigned int pbr_val[4] = { 2, 3, 5, 7 };
+
+	/* Valid baud rate scaler values*/
+	static const unsigned int brs_val[16] = { 2, 4, 6, 8,
+						16, 32, 64, 128,
+						256, 512, 1024, 2048,
+						4096, 8192, 16384, 32768 };
+
+	unsigned int tmp_val = 0, curr_val = 0;
+	unsigned int i = 0, j = 0;
+
+	tmp_val = clkrate / speed_hz;
+
+	for (i = 0; i < ARRAY_SIZE(pbr_val); i++) {
+		for (j = 0; j < ARRAY_SIZE(brs_val); j++) {
+			curr_val = pbr_val[i] * brs_val[j];
+			if (curr_val >= tmp_val) {
+				*req_pbr = i;
+				*req_br = j;
+				return TEE_SUCCESS;
+			}
+		}
+	}
+
+	EMSG("Can not find valid baud rate,speed_hz is %d, ", speed_hz);
+	EMSG("clkrate is %d, we use the max prescaler value.", clkrate);
+
+	return TEE_ERROR_ITEM_NOT_FOUND;
+}
+
+/* setup speed for slave */
+static void dspi_setup_speed(struct nxp_dspi_data *dspi_data,
+			     unsigned int speed)
+{
+	TEE_Result status = TEE_ERROR_GENERIC;
+	unsigned int bus_setup = 0, bus_clock = 0;
+	unsigned int req_i = 0, req_j = 0;
+
+	bus_clock = dspi_data->bus_clk_hz;
+
+	DMSG("DSPI set_speed: expected SCK speed %u, bus_clk %u.", speed,
+	     bus_clock);
+
+	bus_setup = io_read32(dspi_data->base + DSPI_CTAR0);
+	bus_setup &= ~(DSPI_CTAR_BRD | DSPI_CTAR_BRP(0x3) | DSPI_CTAR_BR(0xf));
+
+	status = dspi_convert_hz_to_baud(&req_i, &req_j, speed, bus_clock);
+
+	/* In case of failure scenario with max speed, setting default speed */
+	if (status == TEE_ERROR_ITEM_NOT_FOUND) {
+		speed = dspi_data->speed_hz;
+		EMSG("DSPI set_speed use default SCK rate %u.", speed);
```

Comment:
The error message should be either here or `dspi_convert_hz_to_baud()` but probably not both?